diff --git a/Zend/zend_stream.c b/Zend/zend_stream.c
index 3fd7fa0..a4e9917 100644
--- a/Zend/zend_stream.c
+++ b/Zend/zend_stream.c
@@ -338,6 +338,8 @@ ZEND_API int zend_compare_file_handles(zend_file_handle *fh1, zend_file_handle *
 		return 0;
 	}
 	switch (fh1->type) {
+		case ZEND_HANDLE_FILENAME:
+			return strcmp(fh1->filename, fh2->filename) == 0;
 		case ZEND_HANDLE_FD:
 			return fh1->handle.fd == fh2->handle.fd;
 		case ZEND_HANDLE_FP:
