commit c34895e837b50213c2bb201c612904342d2bd216
Author: Stanislav Malyshev <stas@php.net>
Date:   Mon May 27 18:04:00 2019 -0700

    Fix bug #77967 - Bypassing open_basedir restrictions via file uris

diff --git a/ext/sqlite3/sqlite3.c b/ext/sqlite3/sqlite3.c
index 7e7a3a0454..9e70f06651 100644
--- a/ext/sqlite3/sqlite3.c
+++ b/ext/sqlite3/sqlite3.c
@@ -2034,6 +2034,15 @@ static int php_sqlite3_authorizer(void *autharg, int access_type, const char *ar
 		case SQLITE_ATTACH:
 		{
 			if (memcmp(arg3, ":memory:", sizeof(":memory:")) && *arg3) {
+				if (strncmp(arg3, "file:", 5) == 0) {
+					/* starts with "file:" */
+					if (!arg3[5]) {
+						return SQLITE_DENY;
+					}
+					if (php_check_open_basedir(arg3 + 5)) {
+						return SQLITE_DENY;
+					}
+				}

 #if PHP_API_VERSION < 20100412
 				if (PG(safe_mode) && (!php_checkuid(arg3, NULL, CHECKUID_CHECK_FILE_AND_DIR))) {
