diff --git a/ext/filter/sanitizing_filters.c b/ext/filter/sanitizing_filters.c
index cc13e1a..69ad93d 100644
--- a/ext/filter/sanitizing_filters.c
+++ b/ext/filter/sanitizing_filters.c
@@ -111,7 +111,7 @@ static void php_filter_encode_url(zval *value, const unsigned char* chars, const
 static void php_filter_strip(zval *value, zend_long flags)
 {
 	unsigned char *str;
-	int   i, c;
+	size_t   i, c;
 	zend_string *buf;
 
 	/* Optimization for if no strip flags are set */
@@ -120,7 +120,7 @@ static void php_filter_strip(zval *value, zend_long flags)
 	}
 
 	str = (unsigned char *)Z_STRVAL_P(value);
-	buf = zend_string_alloc(Z_STRLEN_P(value) + 1, 0);
+	buf = zend_string_alloc(Z_STRLEN_P(value), 0);
 	c = 0;
 	for (i = 0; i < Z_STRLEN_P(value); i++) {
 		if ((str[i] >= 127) && (flags & FILTER_FLAG_STRIP_HIGH)) {
@@ -162,7 +162,7 @@ static void filter_map_apply(zval *value, filter_map *map)
 	zend_string *buf;
 
 	str = (unsigned char *)Z_STRVAL_P(value);
-	buf = zend_string_alloc(Z_STRLEN_P(value) + 1, 0);
+	buf = zend_string_alloc(Z_STRLEN_P(value), 0);
 	c = 0;
 	for (i = 0; i < Z_STRLEN_P(value); i++) {
 		if ((*map)[str[i]]) {
