diff --git a/ext/openssl/xp_ssl.c b/ext/openssl/xp_ssl.c
index 193eb53..4f4d6ae 100644
--- a/ext/openssl/xp_ssl.c
+++ b/ext/openssl/xp_ssl.c
@@ -2128,8 +2128,8 @@ static struct timeval subtract_timeval( struct timeval a, struct timeval b )
 	difference.tv_usec = a.tv_usec - b.tv_usec;
 
 	if (a.tv_usec < b.tv_usec) {
-	  	b.tv_sec  -= 1L;
-	   	b.tv_usec += 1000000L;
+	  	difference.tv_sec  -= 1L;
+	   	difference.tv_usec += 1000000L;
 	}
 
 	return difference;
