Index: NEWS
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/NEWS b/NEWS
--- a/NEWS	(revision 058634aba0551f23b3a58f23b421947940e353a6)
+++ b/NEWS	(revision 652e718abbb7cf0e124250aa27c16f6d8e4a8b61)
@@ -184,6 +184,9 @@
   . Fixed bug #76965 (INI_SCANNER_RAW doesn't strip trailing whitespace).
     (Pierrick)
 
+- Standard:
+  . Fixed bug #80366 (Return Value of zend_fstat() not Checked). (sagpant, cmb)
+
 - Tidy:
   . Fixed bug #77027 (tidy::getOptDoc() not available on Windows). (cmb)
 
Index: ext/standard/iptc.c
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/ext/standard/iptc.c b/ext/standard/iptc.c
--- a/ext/standard/iptc.c	(revision 058634aba0551f23b3a58f23b421947940e353a6)
+++ b/ext/standard/iptc.c	(revision 652e718abbb7cf0e124250aa27c16f6d8e4a8b61)
@@ -216,7 +216,9 @@
 	}
 
 	if (spool < 2) {
-		zend_fstat(fileno(fp), &sb);
+		if (zend_fstat(fileno(fp), &sb) != 0) {
+			RETURN_FALSE;
+		}
 
 		spoolbuf = zend_string_safe_alloc(1, iptcdata_len + sizeof(psheader) + 1024 + 1, sb.st_size, 0);
 		poi = (unsigned char*)ZSTR_VAL(spoolbuf);
