diff --git a/ext/standard/image.c b/ext/standard/image.c
index 1d999ca..7447ebd 100644
--- a/ext/standard/image.c
+++ b/ext/standard/image.c
@@ -1474,6 +1474,11 @@ static void php_getimagesize_from_any(INTERNAL_FUNCTION_PARAMETERS, int mode) {
 			return;
 	}
 
+	if (mode == FROM_PATH && CHECK_NULL_PATH(input, input_len)) {
+			php_error_docref(NULL, E_WARNING, "Invalid path");
+			return;
+	}
+
 	if (argc == 2) {
 		zval_dtor(info);
 		array_init(info);
