diff --git a/ext/zip/php_zip.c b/ext/zip/php_zip.c
index aee355b..0659a2f 100644
--- a/ext/zip/php_zip.c
+++ b/ext/zip/php_zip.c
@@ -586,8 +586,9 @@ int php_zip_glob(char *pattern, int pattern_len, zend_long flags, zval *return_v
 		add_next_index_string(return_value, globbuf.gl_pathv[n]+cwd_skip);
 	}
 
+	ret = globbuf.gl_pathc;
 	globfree(&globbuf);
-	return globbuf.gl_pathc;
+	return ret;
 #else
 	php_error_docref(NULL, E_ERROR, "Glob support is not available");
 	return 0;
